core/modules/map.rs

```diff
@@ -1033,21 +1075,56 @@ impl ModuleMap {
     context: v8::Local<'s, v8::Context>,
     specifier: v8::Local<'s, v8::String>,
     _import_attributes: v8::Local<'s, v8::FixedArray>,
-    _referrer: v8::Local<'s, v8::Module>,
+    referrer: v8::Local<'s, v8::Module>,
   ) -> Option<v8::Local<'s, v8::Object>> {
     // SAFETY: `CallbackScope` can be safely constructed from `Local<Context>`
     v8::callback_scope!(unsafe scope, context);
 
-    let specifier_str = specifier.to_rust_string_lossy(scope);
+    let module_map =
+      // SAFETY: We retrieve the pointer from the slot, having just set it a few stack frames up
+      unsafe { scope.get_slot::<*const Self>().unwrap().as_ref().unwrap() };
 
-    let message = v8::String::new(
-      scope,
-      &format!(r#"Module source can not be imported for "{specifier_str}""#),
-    )
-    .unwrap();
-    let exception = v8::Exception::syntax_error(scope, message);
-    scope.throw_exception(exception);
-    None
+    let specifier_str = specifier.to_rust_string_lossy(scope);
+    let referrer_global = v8::Global::new(scope, referrer);
+    let module_reference = {
+      let module_map_data = module_map.data.borrow();
+      let referrer_info = module_map_data
+        .get_info_by_module(&referrer_global)
+        .expect("ModuleInfo not found");
+      let module_request = referrer_info
+        .requests
+        .iter()
+        .find(|r| {
+          r.specifier_key
+            .as_ref()
+            .is_some_and(|s| s == &specifier_str)
+        })
+        .expect("ModuleInfo::requests did not contain a matching specifier_key when getting source");
+      module_request.reference.clone()
+    };
+    match module_map
+      .data
+      .borrow()
+      .sources
+      .get(module_reference.specifier.as_str())
+    {
+      Some(ModuleSourceData {
+        value,
+        kind: ModuleSourceKind::Wasm,
+      }) => Some(v8::Local::new(scope, value)),
+      None => {
+        let message = v8::String::new(
+          scope,
+          &format!(
+            r#"Module source can not be imported for "{specifier_str}""#
```

Comment:
This is a bit hard to read and I was surprised it doesn't cause borrow checker error. Consider rewriting into `if let Some(...) {}`